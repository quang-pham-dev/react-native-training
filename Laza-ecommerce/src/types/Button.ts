import { ImageSourcePropType, ViewStyle } from 'react-native';

export type SocialButtonProps = {
  onPress?: () => void;
  icon?: ImageSourcePropType;
  title: string;
  buttonStyles?: ViewStyle[];
  children?: JSX.Element | JSX.Element[];
  disabled?: boolean;
};

export type BottomButtonProps = {
  onPress?: () => void;
  title: string;
  buttonStyles?: ViewStyle[];
  children?: JSX.Element | JSX.Element[];
  disabled?: boolean;
};

export type GoBackButtonProps = {
  onPress?: () => void;
  icon: ImageSourcePropType;
  buttonStyles?: ViewStyle[];
  disabled?: boolean;
};
