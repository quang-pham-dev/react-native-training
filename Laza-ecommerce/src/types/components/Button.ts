import { ImageSourcePropType, StyleProp, TextStyle, ViewStyle } from 'react-native';

export interface IButtonProps {
  testID?: string;
  onPressHandler?: () => void;
  buttonStyles?: ViewStyle[];
  text: string;
  textStyles?: StyleProp<TextStyle>;
  icon?: ImageSourcePropType;
  iconStyles?: ViewStyle[];
  type?: string;
  disabled?: boolean;
  color?: string;
  children?: JSX.Element | JSX.Element[];
}
