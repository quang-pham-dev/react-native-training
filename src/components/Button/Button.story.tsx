import React from 'react'
import {storiesOf} from '@storybook/react-native'
import {select, text} from '@storybook/addon-knobs'
import {action} from '@storybook/addon-actions'

import {StoryScreen, Story, UseCase} from '../../../storybook/views'
import {Button, SocialButton, BtnType} from './index'
import {Colors, Icons, Metrics} from '@themes'
import ViewStyled from '@components/View/View.styles'

const typesOption = Object.values(BtnType)

declare let module: NodeModule

storiesOf('Button', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Bottom', () => (
    <Story>
      <UseCase text="Button with type bottom">
        <Button
          onPress={action('Pressed!')}
          type={select('type', typesOption, typesOption[0])}
          label={text('text', 'Title for Button')}
        />
      </UseCase>
    </Story>
  ))
  .add('Social', () => (
    <Story>
      <UseCase text="Button with type social">
        <SocialButton
          type={select('type', typesOption, typesOption[1])}
          icon={Icons.google}
          bgColor={Colors.palette.google}
          onPress={action('Pressed!')}
          label={text('text', 'Google')}
        />
        <ViewStyled.Custom mVertical={Metrics.margin.small}>
          <SocialButton
            type={select('type', typesOption, typesOption[1])}
            icon={Icons.facebook}
            bgColor={Colors.palette.facebook}
            onPress={action('Pressed!')}
            label={text('text', 'Facebook')}
          />
        </ViewStyled.Custom>
        <SocialButton
          type={select('type', typesOption, typesOption[1])}
          icon={Icons.twitter}
          bgColor={Colors.palette.twitter}
          onPress={action('Pressed!')}
          label={text('text', 'Twitter')}
        />
      </UseCase>
    </Story>
  ))
  .add('Disable Button', () => (
    <Story>
      <UseCase text="Button disabled">
        <Button
          onPress={action('Pressed!')}
          type={select('type', typesOption, typesOption[0])}
          label={text('text', 'Title for Button')}
          disabled
        />
      </UseCase>
    </Story>
  ))
  .add('Loading Button', () => (
    <Story>
      <UseCase text="Button loading">
        <Button
          onPress={action('Pressed!')}
          type={select('type', typesOption, typesOption[0])}
          label={text('text', 'Title for Button')}
          isLoading
        />
      </UseCase>
    </Story>
  ))
