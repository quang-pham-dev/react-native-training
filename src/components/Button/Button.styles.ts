import styled from 'styled-components/native'

// Services
import {getType, convertToPx} from '@services/format'

// Themes
import {Colors, Metrics} from '@themes'

type ButtonProps = {
  type: string
  bgColor?: string
  alignItems?: string
  justifyContent?: string
  disabled?: boolean
  children: React.ReactNode
  pTop?: number
}

const element = 'button'

export const ButtonWrapperStyled = styled.TouchableHighlight<ButtonProps>`
  align-items: ${props => props.alignItems || 'center'};
  padding-top: ${props => props.pTop || Metrics.padding.mediumPlus}px
    ${props => {
      const type = getType(props.type, element)

      if (type) {
        const {
          width,
          height,
          borderRadius,
          bgColor,
          borderColor,
          disableBgColor,
          disableBorderColor,
        } = type
        return `
        ${width ? `width: ${convertToPx(width)};` : ''}
        ${height ? `height: ${convertToPx(height)};` : ''}
        ${borderRadius ? `border-radius: ${convertToPx(borderRadius)};` : ''}
        ${
          bgColor
            ? `background-color: ${props.bgColor ? disableBgColor : bgColor};`
            : ''
        }
        ${
          borderColor
            ? `border: 1px solid ${
                props.disabled ? disableBorderColor : borderColor
              };`
            : ''
        }
      `
      }
    }};
`

export const ButtonLabelStyled = styled.Text<ButtonProps>`
  ${props => {
    const type = getType(props.type, element)

    if (type) {
      const {fontSize, fontFamily, color, disableColor} = type
      return `
        ${fontSize ? `font-size: ${convertToPx(fontSize)};` : ''}
        ${fontFamily ? `font-family: ${fontFamily};` : ''}
        ${color ? `color: ${props.disabled ? disableColor : color};` : ''}
      `
    }
  }};
`

export const ButtonSocialWrapperStyled = styled.TouchableHighlight<ButtonProps>`
  align-items: ${props => props.alignItems || 'center'};
  justify-content: ${props => props.justifyContent || 'center'};
  background-color: ${props => props.bgColor || Colors.palette.transparent};
  ${props => {
    const type = getType(props.type, element)

    if (type) {
      const {
        width,
        height,
        borderRadius,
        bgColor,
        borderColor,
        disableBgColor,
        disableBorderColor,
      } = type
      return `
        ${width ? `width: ${convertToPx(width)};` : ''}
        ${height ? `height: ${convertToPx(height)};` : ''}
        ${borderRadius ? `border-radius: ${convertToPx(borderRadius)};` : ''}
        ${
          bgColor
            ? `background-color: ${props.bgColor ? disableBgColor : bgColor};`
            : ''
        }
        ${
          borderColor
            ? `border: 1px solid ${
                props.disabled ? disableBorderColor : borderColor
              };`
            : ''
        }
        `
    }
  }};
`
