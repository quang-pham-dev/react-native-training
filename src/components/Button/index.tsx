import React from 'react'
import {ImageSourcePropType} from 'react-native'

// LIBS
import {s} from 'react-native-size-matters/extend'

// Components
import LoadingIndicator, {LoadingSize} from '@components/LoadingIndicator'

// Styles
import IconStyled from '@components/Icon/Icon.styles'
import FlexStyled from '@components/Flex/Flex.styles'
import {
  ButtonLabelStyled,
  ButtonSocialWrapperStyled,
  ButtonWrapperStyled,
} from './Button.styles'

// Themes
import {Colors} from '@themes'

export enum BtnType {
  BOTTOM = 'bottom',
  SOCIAL = 'social',
}

type ButtonProps = {
  type: BtnType
  label: string
  onPress: () => void
  disabled?: boolean
  isLoading?: boolean
  btnProps?: {
    [key: string]: string | number
  }
}

const typeActive = (type: BtnType) =>
  [BtnType.BOTTOM, BtnType.SOCIAL].includes(type)
    ? Colors.palette.lightGray
    : Colors.palette.primary

const Button = ({
  type = BtnType.BOTTOM,
  label,
  onPress,
  disabled,
  isLoading,
  btnProps,
}: ButtonProps) => {
  return (
    <ButtonWrapperStyled
      type={type}
      onPress={disabled ? undefined : onPress}
      disabled={disabled}
      underlayColor={typeActive(type)}
      {...btnProps}>
      <>
        {isLoading && (
          <LoadingIndicator
            size={LoadingSize.SMALL}
            color={Colors.palette.white}
          />
        )}
        {!isLoading && label && (
          <ButtonLabelStyled disabled={disabled} type={type}>
            {label}
          </ButtonLabelStyled>
        )}
      </>
    </ButtonWrapperStyled>
  )
}

type SocialButtonProps = {
  label: string
  type: BtnType
  disabled?: boolean
  onPress: () => void
  bgColor?: string
  isLoading?: boolean
  icon?: ImageSourcePropType
  btnProps?: {
    [key: string]: string | number
  }
}

const SocialButton = ({
  type = BtnType.SOCIAL,
  label,
  onPress,
  isLoading,
  disabled,
  bgColor,
  icon,
  btnProps,
}: SocialButtonProps) => {
  return (
    <ButtonSocialWrapperStyled
      type={type}
      bgColor={bgColor}
      onPress={disabled ? undefined : onPress}
      disabled={disabled}
      underlayColor={typeActive(type)}
      {...btnProps}>
      <>
        {isLoading && (
          <LoadingIndicator
            size={LoadingSize.SMALL}
            color={Colors.palette.white}
          />
        )}
        {!isLoading && (
          <FlexStyled.RowCenterHorizontal>
            {icon && (
              <FlexStyled.Custom pRight={s(4)}>
                <IconStyled width={s(16)} height={s(16)} source={icon} />
              </FlexStyled.Custom>
            )}
            {label && (
              <ButtonLabelStyled disabled={disabled} type={type}>
                {label}
              </ButtonLabelStyled>
            )}
          </FlexStyled.RowCenterHorizontal>
        )}
      </>
    </ButtonSocialWrapperStyled>
  )
}

export {Button, SocialButton}
