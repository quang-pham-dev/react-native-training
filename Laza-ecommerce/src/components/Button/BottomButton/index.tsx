import React, { memo } from 'react';
import { Text, TouchableOpacity } from 'react-native';
// Types
import { BottomButtonProps } from 'types/Button';
// Styles
import styles from './styles';

const BottomButton = ({
  title,
  disabled,
  buttonStyles = [],
  children = [],
  onPress = () => {},
}: BottomButtonProps) => {
  return (
    <TouchableOpacity
      style={[styles.button, buttonStyles]}
      onPress={onPress}
      disabled={disabled}
      activeOpacity={0.8}
    >
      <Text style={[styles.title]}>{title}</Text>
      {children}
    </TouchableOpacity>
  );
};

export default memo(BottomButton);
