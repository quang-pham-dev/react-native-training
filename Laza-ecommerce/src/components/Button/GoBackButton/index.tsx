import React, { memo } from 'react';
import { Image, TouchableOpacity } from 'react-native';
// Types
import { GoBackButtonProps } from 'types/Button';
// Styles
import styles from './styles';

const GoBackButton = ({
  icon,
  disabled,
  buttonStyles = [],
  onPress = () => {},
}: GoBackButtonProps) => {
  return (
    <TouchableOpacity
      style={[styles.button, buttonStyles]}
      onPress={onPress}
      disabled={disabled}
      activeOpacity={0.8}
    >
      {icon && <Image style={styles.icon} source={icon} />}
    </TouchableOpacity>
  );
};

export default memo(GoBackButton);
