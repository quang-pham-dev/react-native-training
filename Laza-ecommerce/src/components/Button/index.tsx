import React from 'react';
import { Image, Text, TouchableOpacity } from 'react-native';
// Types
import { ButtonProps } from 'types/Button';
// Styles
import styles from './styles';

const Button = ({
  onPress = () => {},
  text,
  textStyles,
  buttonStyles,
  icon,
  iconStyles,
  disabled,
  children,
  ...props
}: ButtonProps) => {
  return (
    <TouchableOpacity
      style={[styles.button, buttonStyles]}
      {...props}
      onPress={onPress}
      disabled={disabled}
      activeOpacity={0.8}
    >
      {icon && <Image style={[styles.icon, iconStyles]} source={icon} />}
      {text && <Text style={[styles.text , textStyles]}>{text}</Text>}
      {children}
    </TouchableOpacity>
  );
};

export default Button;
