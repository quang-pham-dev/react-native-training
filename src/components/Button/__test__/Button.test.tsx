import React from 'react'

// Libs
import {fireEvent, render} from '@testing-library/react-native'

// Components
import {Button, SocialButton, BtnType} from '@components/Button'

const props = {
  onPress: jest.fn(),
  type: BtnType.BOTTOM,
  label: 'Title Button',
  disabled: false,
  isLoading: false,
}

const customProps = {
  disabled: true,
  isLoading: true,
  onPress: jest.fn(),
  label: 'Btn',
  type: BtnType.SOCIAL,
}

describe('Button', () => {
  test('Should render correctly', () => {
    const component = render(<Button {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should call function onPress', () => {
    const {getByText} = render(<Button {...props} />)
    const button = getByText('Title Button')
    fireEvent.press(button)
    expect(props.onPress).toHaveBeenCalled()
  })

  test('Should render with custom props', () => {
    const element = render(<Button {...customProps} />)
    const tree = element.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
describe('SocialButton', () => {
  test('Should render correctly', () => {
    const component = render(<SocialButton {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should call function onPress', () => {
    const {getByText} = render(<SocialButton {...props} />)
    const button = getByText('Title Button')
    fireEvent.press(button)
    expect(props.onPress).toHaveBeenCalled()
  })

  test('Should render with custom props', () => {
    const element = render(<SocialButton {...customProps} />)
    const tree = element.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
