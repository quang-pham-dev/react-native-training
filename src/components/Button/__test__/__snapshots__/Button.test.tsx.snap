// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should render correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#3D93F8",
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderWidth": 1,
        "height": 75,
        "paddingTop": 20,
        "width": "auto",
      },
    ]
  }
>
  <Text
    disabled={false}
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Inter-Medium",
          "fontSize": 17,
        },
      ]
    }
    type="bottom"
  >
    Title Button
  </Text>
</View>
`;

exports[`Button Should render with custom props 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "transparent",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "height": 50,
        "paddingTop": 20,
        "width": "auto",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <ActivityIndicator
      color="#FFFFFF"
      size="small"
    />
  </View>
</View>
`;

exports[`SocialButton Should render correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#3D93F8",
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderWidth": 1,
        "height": 75,
        "justifyContent": "center",
        "width": "auto",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ]
    }
  >
    <Text
      disabled={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Inter-Medium",
            "fontSize": 17,
          },
        ]
      }
      type="bottom"
    >
      Title Button
    </Text>
  </View>
</View>
`;

exports[`SocialButton Should render with custom props 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "height": 50,
        "justifyContent": "center",
        "width": "auto",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "width": "100%",
        },
      ]
    }
  >
    <ActivityIndicator
      color="#FFFFFF"
      size="small"
    />
  </View>
</View>
`;
